manager/common/license.go

```diff
@@ -0,0 +1,192 @@
+	c.queue[vendor] = c.queue[vendor][1:]
+	delete(c.clients, key)
+	log.Printf("Client %s removed from the %s queue \n", key, vendor)
+}
+
+func (c *LicenseCounter) del(key string, vendor string) int {
+	index := -1
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	for i := 0; i < len(c.queue[vendor]); i++ {
+		if c.queue[vendor][i] == key {
+			c.queue[vendor] = append(c.queue[vendor][:i], c.queue[vendor][i+1:]...)
+			delete(c.clients, key)
+			log.Printf("Job cancelled by client %s - removed from queue \n", key)
+			index = i
+			break
+		}
+	}
+	if index == -1 {
+		log.Fatalf("%s is not in queue \n", key)
+	}
+	return index
+}
+
+func (c *LicenseCounter) peekQueue(vendor string, hash string) bool {
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	return c.queue[vendor][0] == hash
+}
+
+type Client struct {
+	IP       string
+	User     string
+	Start    time.Time
+	Quantity int32
+	Vendor   string
+	Feature  string
+}
+
+type License struct {
+	Total int
+	Used  int
+}
+
+var clients = map[string]*Client{}
+var totalCadenceLic = 3
+var totalXilinxLic = 3
+var licenseCounter = LicenseCounter{
+	licenses: map[string]*License{"xilinx": &License{Used: 0, Total: totalXilinxLic}, "cadence": &License{Used: 0, Total: totalCadenceLic}},
+	clients:  make(map[string]*Client),
+	queue:    map[string][]string{"xilinx": make([]string, 0), "cadence": make([]string, 0)}}
+
+type Server struct {
+	rng *rand.Rand
+}
+
+func (s *Server) Polling(stream rpc_license.License_PollingServer) error {
+	var hash string
+	var vendor string
+	s.rng = rand.New(rand.NewSource(rand.Int63()))
+	for {
+		recv, err := stream.Recv()
+		// remove client hash from queue when connection is broken
+		if err == io.EOF {
+			licenseCounter.del(hash, vendor)
+			return nil
+		}
+		if err != nil {
+			licenseCounter.del(hash, vendor)
+			return err
+		}
+		md, ok := metadata.FromIncomingContext(stream.Context())
+		if !ok {
+			log.Printf("grpc metadata not ok \n")
+		}
+		ip := strings.Join(md[":authority"], " ")
+		username := recv.User
+		start := time.Now().UTC()
+		vendor = recv.Vendor
+		if recv.Hash == "" {
+			hash, err = astoreServer.GenerateUid(s.rng)
+			if err != nil {
+				log.Printf("Failed to generate hash for %s@%s", recv.User, ip)
+				return err
+			}
+			client := Client{IP: ip, User: username, Start: start, Quantity: recv.Quantity, Vendor: vendor, Feature: recv.Feature}
+			licenseCounter.enqueue(hash, &client)
+		} else {
+			hash = recv.Hash
+		}
+		_, total, available := licenseCounter.peekLicense(vendor)
+		if available && licenseCounter.peekQueue(vendor, hash) {
+			log.Printf("%s %s licenses in use: %d/%d \n", vendor, recv.Feature,
+				licenseCounter.increment(vendor, int(recv.Quantity), hash),
+				total)
+			licenseCounter.dequeue(hash, vendor)
+			err := stream.Send(&rpc_license.PollingResponse{Acquired: true, Hash: hash})
+			if err != nil {
+				log.Printf("Failed to send message back to client %s: %s \n", hash, err)
+			} else {
+				break
```

Comment:
should the break happen error or not? don't think the stream is valid after a failed send? would need to check the docs. Generally on send or receive error the stream is closed / client dropped, there's clearly a communication problem.